Update documentation for boxes_flow, allow None

jstockwin opened the pull request:
**Pull request**

Closes #395 

- Allows passing `boxes_flow` as `None` to disable the advanced layout analysis
- Updates the documentation to this effect (if you want me to call it something other than "advanced layout analysis - let me know!)
- I've added a private `__validate` method to `LAParams` to check that boxes flow is either `None`, or between `-1` and `+1`. Happy to remove this if you think it's excessive..

- I've NOT added any other checks about any other params to `__validate`. I'm happy to do this, but there don't seem to be any other specified ranges in the [docs](https://pdfminersix.readthedocs.io/en/latest/api/composable.html#laparams), so I'd just be checking the types, really.

**How Has This Been Tested?**

I checked before the change that passing `None` breaks, and after that it works (using `extract_pages`). The speed changes a lot without the extra layout analysis, so I could tell it was working. `tox` passes.

**Checklist**

- [X] I have added tests that prove my fix is effective or that my feature 
  works
- [X] I have added docstrings to newly created methods and classes
- [X] I have optimized the code at least one time after creating the initial 
  version
- [X] I have updated the [README.md](../README.md) or I am verified that this
  is not necessary
- [X] I have updated the [readthedocs](../docs/source) documentation or I 
  verified that this is not necessary
- [X] I have added a consice human-readable description of the change to 
  [CHANGELOG.md](../CHANGELOG.md)

pietermarsman commented on CHANGELOG.md:
```diff
@@ -8,6 +8,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ### Fixed
 - Grouping of text lines outside of parent container bounding box ([#386](https://github.com/pdfminer/pdfminer.six/pull/386))
 
+### Changed
```
I think this is added functionality. It depends on how you define "functionality" though. As I see it, all the documented input/output remain the same, we just *add* a new option to pass `None`.

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -72,8 +78,21 @@ def __init__(self,
+                    isinstance(self.boxes_flow, float)):
+                raise InvalidLaParamsException(boxes_flow_err_msg)
+            if not -1 <= self.boxes_flow <= 1:
+                raise InvalidLaParamsException(boxes_flow_err_msg)
```
Use `ValueError` for this. 

From the documentation of `ValueError`:

> Raised when an operation or function receives an argument that has the right type but an inappropriate value

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -72,8 +78,21 @@ def __init__(self,
+                                  "number between -1 and +1")
+            if not (isinstance(self.boxes_flow, int) or
+                    isinstance(self.boxes_flow, float)):
+                raise InvalidLaParamsException(boxes_flow_err_msg)
```
Use `TypeError` for this. 

From the documentation of `TypeError`:

> Passing arguments of the wrong type (e.g. passing a list when an int is expected) should result in a TypeError

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -50,7 +54,9 @@ class LAParams:
         should be within the range of -1.0 (only horizontal position
-        matters) to +1.0 (only vertical position matters).
+        matters) to +1.0 (only vertical position matters). You can also pass
+        None to disable advanced layout analysis, and instead return text
```
I think you can use \`None\`, such that Spinx renders it as code.

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -72,8 +78,21 @@ def __init__(self,
+        self.__validate()
         return
 
+    def __validate(self):
```
Use single underscore

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -723,8 +742,8 @@ def analyze(self, laparams):
-        if -1 <= laparams.boxes_flow and laparams.boxes_flow <= +1 \
-                and textboxes:
+        if laparams.boxes_flow is not None and \
+                -1 <= laparams.boxes_flow <= +1 and textboxes:
```
This can be more simple:

```python
if laparams.boxes_flow is None:
    ....
elif -1 <- laparams.boxes_flow <= 1 and textboxes:
    ...
else:
    # should never happend due to _validate()
```

You can even do:

```python
if laparams.boxes_flow is None:
    ...
else:
    ...
```

if both sorting methods also work on an empty list.

jstockwin commented:
@pietermarsman Thanks for the review - all makes sense.
I've rebased the first commit to resolve conflicts in `CONTRIBUTING.MD`, and then added a new commit which addresses your comments.
Please take another look when you get a chance.

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -50,7 +50,9 @@ class LAParams:
-        matters) to +1.0 (only vertical position matters).
+        matters) to +1.0 (only vertical position matters). You can also pass
+        `None` to disable advanced layout analysis, and instead return text
+        based on the (x, y) coordinates of the bottom left corner.
```
> based on the (x, y) coordinates of the bottom left corner

Took me 3 times to understand that it is the "bottom left corner" _of the text boxes_.

jstockwin replied:
I'll re-word it :)

pietermarsman commented on pdfminer/layout.py:
```diff
@@ -72,8 +74,21 @@ def __init__(self,
         return
 
+    def _validate(self):
+        # Validate boxes_flow
```
Use docstring with `"""`

jstockwin replied:
Hm, this was actually intended as a "this part of the code is validating the boxes_flow parameter", since perhaps in the future we will validate more things.

I personally think it's silly to add a docstring to a function called `_validate` which would basically be just `"""validate the params"""`, so if it works for you I think I'll just remove this comment?

jstockwin commented:
I've just noticed there are a few other places that the documentation needs updating - don't merge until I've fixed that :)

jstockwin commented:
Okay, now this should be good :+1:

pietermarsman commented:
Wow: one of the tests failed because apparently tox doesn't work with python 3.4 anymore. In this PR we can solve it by specifying an older version of tox. 

I'll see if we can drop support for python 3.4 in the future.

jstockwin commented:
Agree, except I think this project depends on `tox_travis` which just installed `tox>=2`.

This installed the latest version of tox, which is v3.14.6. Tox deprecated Python 3.4 in v3.14.1, but the deprecation says it won't actually get removed until the next major release. (tox [changelog](https://tox.readthedocs.io/en/latest/changelog.html))

I don't think specifying a specific version of `tox_travis` will fix this, as it's the version of tox that matters... I guess I'll try pinning `tox` to a specific version and hopefully that'll help? Really I think these are issues for `tox_travis`/`tox`, though.

That said, I agree dropping python 3.4 support is a good idea.

jstockwin commented:
Okay, that worked - tests are now green :heavy_check_mark:
